### PR TITLE
[Contracts/Deprecation] don't use assert(), rename to trigger_deprecation()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR is a follow up the discussion that happened in #35526.

I applied all the suggestions I received so far on this previous PR so that we can discuss them all together.

Here are the changes:
- the most important change is to *not* use `assert()` anymore. This fixes the objection from @ostrolucky and @derrabus that deprecations and assert() should not be bound to each other. Instead, in order to still make the function a no-op when wanted, it is suggested that apps declare an empty function. This will be more flexible in practice I believe and keeps all the benefits we envisioned by using assert();
- as suggested by @fabpot, the function is renamed `trigger_deprecation()` instead of `deprecated()`. Because the implementation is now not conditional anymore (from the package pov, since assert() is not used), my arguments to keep the previous name don't stand as strong as before to me so I'm fine renaming.
- type hints are added back (requiring PHP 7.1 to use `void`). As mentioned in the 1st point, ppl that really want deprecations to be no-ops will redeclare the function as empty, thus not using any types, thus reclaiming all the perf diff. And for ppl that do want to catch deprecations, the overhead of types is negligible compared to any processing of the deprecations themselves.

WDYT?

All points can be discussed separately if needed of course. I'm personally fine with merging the PR as is, with all 3 changes.

### CODE REVIEW

File: src/Symfony/Contracts/Deprecation/README.md [hunk 1]
```diff
@@ -15,8 +16,11 @@ The function requires at least 3 arguments:
 
 Example:
 ```php
-deprecated('symfony/blockchain', 8.9, 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
+trigger_deprecation('symfony/blockchain', 8.9, 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
```
Comment (alcaeus) [inline src/Symfony/Contracts/Deprecation/README.md]: One problem has been brought up that affects the version: locale-aware casts from float to string. This example should pass the version on string. For this effect, I would show people to not rely on it until the locale-awareness on casts has been removed in PHP.

**Review by alcaeus — CHANGES_REQUESTED** [review]
The example needs adjustment to remove the implicit cast. The rest of this is a great improvement in the implementation!

File: src/Symfony/Contracts/Deprecation/README.md (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/README.md]: updated.

**Review by alcaeus — APPROVED** [review]
Looks great!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you for listening :)

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Contracts/Deprecation/README.md [hunk 2]
```diff
@@ -15,8 +16,11 @@ The function requires at least 3 arguments:
...
 Example:
 ```php
-deprecated('symfony/blockchain', 8.9, 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
+trigger_deprecation('symfony/blockchain', '8.9', 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
 ```
 
 This will generate the following message:
 `Since symfony/blockchain 8.9: Using "bitcoin" is deprecated, use "fabcoin" instead.`
+
+While no necessarily recommended, the deprecation notices can be completely ignored by declaring an empty
```
Comment (jdreesen) [inline src/Symfony/Contracts/Deprecation/README.md]:
```suggestion
While not necessarily recommended, the deprecation notices can be completely ignored by declaring an empty
```

File: src/Symfony/Contracts/Deprecation/README.md [hunk 3]
```diff
@@ -15,8 +16,11 @@ The function requires at least 3 arguments:
...
+`function trigger_deprecation() {}` in you application.
```
Comment (jdreesen) [inline src/Symfony/Contracts/Deprecation/README.md]:
```suggestion
`function trigger_deprecation() {}` in your application.
```

**Review by jdreesen — CHANGES_REQUESTED** [review]

File: src/Symfony/Contracts/Deprecation/README.md (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/README.md]: fixed thanks

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by derrabus** [comment]
Thank you very much! 😃

**Pull request merged**